File: src/pinky2/pinky2/auto_parking_node.py
#!/usr/bin/env python3
import math
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy
from nav2_msgs.action import NavigateToPose
from geometry_msgs.msg import PoseStamped, Twist, PoseWithCovarianceStamped
from std_msgs.msg import UInt16MultiArray, Bool
from sensor_msgs.msg import Range

AMCL_QOS = QoSProfile(
    depth=1,
    reliability=ReliabilityPolicy.RELIABLE,
    durability=DurabilityPolicy.TRANSIENT_LOCAL,
)

WHITE_THRESHOLD = 2000
LINEAR_VEL  = 0.05
ANGULAR_VEL = 0.5
ROTATION_YAW_TOLERANCE = math.radians(3)
US_STOP_DIST = 0.065

LINE_KP = -0.0005
LINE_ANGULAR_LIMIT = 0.3

LEFT  = 0
MID   = 1
RIGHT = 2


def quat_to_yaw(q) -> float:
    return math.atan2(2.0 * (q.w * q.z + q.x * q.y),
                      1.0 - 2.0 * (q.y * q.y + q.z * q.z))


class AutoParkingNode(Node):
    def __init__(self):
        super().__init__('auto_parking_node')

        #self.declare_parameter('goal_x', 0.001)
        #self.declare_parameter('goal_y', -0.005)
        self.declare_parameter('goal_x', 0.0)
        self.declare_parameter('goal_y', 0.0)
        self.declare_parameter('white_threshold', WHITE_THRESHOLD)
        self.declare_parameter('us_stop_dist', US_STOP_DIST)
        self.declare_parameter('line_kp', LINE_KP)
        self.declare_parameter('line_angular_limit', LINE_ANGULAR_LIMIT)

        self.goal_x = self.get_parameter('goal_x').value
        self.goal_y = self.get_parameter('goal_y').value
        self.white_threshold = self.get_parameter('white_threshold').value
        self.us_stop_dist = self.get_parameter('us_stop_dist').value
        self.line_kp = self.get_parameter('line_kp').value
        self.line_angular_limit = self.get_parameter('line_angular_limit').value

        self.started    = False
        self.ir_active  = False
        self.us_active  = False
        self.rotating   = False
        self.done       = False

        self.current_yaw = None
        self.target_yaw  = None

        self.cmd_pub    = self.create_publisher(Twist, '/pinky1/cmd_vel', 10)
        self.status_pub = self.create_publisher(Bool, '/pinky1/parking/auto_done', 10)
        self._nav_client = ActionClient(self, NavigateToPose, '/pinky1/navigate_to_pose')

        self.create_subscription(
            PoseWithCovarianceStamped, '/pinky1/amcl_pose', self._pose_callback, AMCL_QOS
        )
        self.create_subscription(UInt16MultiArray, '/pinky1/ir_sensor/range', self._ir_callback, 10)
        self.create_subscription(Range, '/pinky1/us_sensor/range', self._us_callback, 10)
        self.create_subscription(Bool, '/pinky1/parking/auto_start', self._start_callback, 10)

        self.create_timer(0.1, self._rotation_tick)

        # wait_for_server 블로킹 방지 → 타이머로 Nav2 서버 준비 확인 후 시작
        self._nav_check_timer = self.create_timer(1.0, self._check_nav_server)

        self.get_logger().info(
            f'AutoParkingNode 시작. 목표=({self.goal_x}, {self.goal_y})'
        )
        self._go_home()
    def _check_nav_server(self):
        """Nav2 서버가 준비되면 자동으로 시작"""
        if self._nav_client.server_is_ready():
            self.get_logger().info('Nav2 서버 준비 완료. 이동 시작.')
            self.destroy_timer(self._nav_check_timer)
            self.started = True
            self._go_home()

    def _start_callback(self, msg: Bool):
        if msg.data and not self.started:
            self.started = True
            self.get_logger().info('주차 진입 시작 신호 수신.')
            self._go_home()

    def _finish(self, success: bool):
        self.status_pub.publish(Bool(data=success))
        self.done = True
        # 타이머로 약간 딜레이 후 shutdown (메시지 전송 여유)
        self.create_timer(0.3, self._shutdown)

    def _shutdown(self):
        rclpy.shutdown()

    # ── Phase 1: 네비게이션 ──────────────────────────────────────

    def _go_home(self):
        goal = NavigateToPose.Goal()
        goal.pose = PoseStamped()
        goal.pose.header.frame_id = 'map'
        goal.pose.header.stamp = self.get_clock().now().to_msg()
        goal.pose.pose.position.x = self.goal_x
        goal.pose.pose.position.y = self.goal_y

        self.get_logger().info(f'목표 전송: ({self.goal_x}, {self.goal_y})')
        future = self._nav_client.send_goal_async(
            goal, feedback_callback=self._nav_feedback
        )
        future.add_done_callback(self._nav_goal_response)

    def _nav_feedback(self, feedback_msg: NavigateToPose.Feedback):
        dist = feedback_msg.feedback.distance_remaining
        self.get_logger().info(f'이동 중... 남은 거리: {dist:.2f} m', throttle_duration_sec=1.0)

    def _nav_goal_response(self, future):
        handle = future.result()
        if not handle.accepted:
            self.get_logger().error('목표 거부됨. 종료.')
            self._finish(success=False)
            return
        self.get_logger().info('목표 수락됨.')
        handle.get_result_async().add_done_callback(self._nav_result)

    def _nav_result(self, future):
        status = future.result().status
        if status == 4:
            self.get_logger().info('도착. IR 주차 모드 시작.')
            self._start_ir_parking()
        else:
            self.get_logger().error(f'네비게이션 실패 (status={status}). 종료.')
            self._finish(success=False)

    # ── Phase 2: IR 흰선 감지 ────────────────────────────────────

    def _start_ir_parking(self):
        self.ir_active = True

    def _ir_callback(self, msg: UInt16MultiArray):
        if not self.ir_active:
            return

        data = msg.data
        left_raw, mid_raw, right_raw = data[LEFT], data[MID], data[RIGHT]

        if mid_raw <= self.white_threshold:
            self.cmd_pub.publish(Twist())
            self.ir_active = False
            self.get_logger().info(
                f'흰선 감지 완료. 초음파 정밀 주차 시작. (L={left_raw}, M={mid_raw}, R={right_raw})'
            )
            self._start_us_parking()
            return

        error = float(left_raw - right_raw)
        angular_z = self.line_kp * error
        angular_z = max(-self.line_angular_limit, min(self.line_angular_limit, angular_z))

        self.get_logger().info(
            f'라인트레이싱: L={left_raw}, M={mid_raw}, R={right_raw}, '
            f'error={error:.0f}, angular_z={angular_z:.3f}',
            throttle_duration_sec=0.5,
        )

        twist = Twist()
        twist.linear.x = LINEAR_VEL
        twist.angular.z = angular_z
        self.cmd_pub.publish(twist)

    # ── Phase 3: 초음파 정밀 주차 ────────────────────────────────

    def _start_us_parking(self):
        self.us_active = True

    def _us_callback(self, msg: Range):
        if not self.us_active:
            return

        dist = msg.range
        self.get_logger().info(f'초음파 거리: {dist:.3f} m', throttle_duration_sec=1.0)

        if dist <= self.us_stop_dist:
            self.cmd_pub.publish(Twist())
            self.us_active = False
            self.get_logger().info(f'정밀 주차 완료 — 장애물까지 {dist:.3f} m. 180도 회전 시작.')
            self._start_rotation()
            return

        twist = Twist()
        twist.linear.x = LINEAR_VEL
        self.cmd_pub.publish(twist)

    # ── amcl_pose 수신 ───────────────────────────────────────────

    def _pose_callback(self, msg: PoseWithCovarianceStamped):
        self.current_yaw = quat_to_yaw(msg.pose.pose.orientation)

    # ── Phase 4: 180도 회전 ──────────────────────────────────────

    def _start_rotation(self):
        self.rotating = True
        self.target_yaw = None

    def _rotation_tick(self):
        if not self.rotating or self.done:
            return
        if self.current_yaw is None:
            return

        if self.target_yaw is None:
            self.target_yaw = self.current_yaw + math.pi
            if self.target_yaw > math.pi:
                self.target_yaw -= 2 * math.pi
            elif self.target_yaw < -math.pi:
                self.target_yaw += 2 * math.pi
            self.get_logger().info(f'회전 목표 yaw: {math.degrees(self.target_yaw):.1f}°')

        angle_err = self.target_yaw - self.current_yaw
        if angle_err > math.pi:
            angle_err -= 2 * math.pi
        elif angle_err < -math.pi:
            angle_err += 2 * math.pi

        self.get_logger().info(
            f'회전 중... 현재={math.degrees(self.current_yaw):.1f}°, '
            f'목표={math.degrees(self.target_yaw):.1f}°, 오차={math.degrees(angle_err):.1f}°',
            throttle_duration_sec=0.5,
        )

        if abs(angle_err) <= ROTATION_YAW_TOLERANCE:
            self.cmd_pub.publish(Twist())
            self.rotating = False
            self.get_logger().info('180도 회전 완료. 시나리오 종료.')
            self._finish(success=True)
            return

        angular_z = max(-ANGULAR_VEL, min(ANGULAR_VEL, ANGULAR_VEL * angle_err))
        twist = Twist()
        twist.angular.z = angular_z
        self.cmd_pub.publish(twist)


def main(args=None):
    rclpy.init(args=args)
    node = AutoParkingNode()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: src/pinky2_cpp/src/auto_parking_node.cpp
#include <cmath>
#include <chrono>
#include <functional>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "nav2_msgs/action/navigate_to_pose.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "geometry_msgs/msg/pose_with_covariance_stamped.hpp"
#include "std_msgs/msg/bool.hpp"
#include "std_msgs/msg/u_int16_multi_array.hpp"
#include "sensor_msgs/msg/range.hpp"

using namespace std::chrono_literals;  // 100ms, 1s 같은 표현 사용 가능
using NavigateToPose = nav2_msgs::action::NavigateToPose;
using GoalHandleNav = rclcpp_action::ClientGoalHandle<NavigateToPose>;

class AutoParkingNode : public rclcpp::Node
{
public:
    // =================================================
    // Python: def __init__(self):
    // C++:    생성자
    // =================================================
    AutoParkingNode() : Node("auto_parking_node")
    {
        // ----- 파라미터 선언 -----
        // Python: self.declare_parameter('goal_x', 0.0)
        // C++:    this->declare_parameter("goal_x", 0.0);
        this->declare_parameter("goal_x", 0.0);
        this->declare_parameter("goal_y", 0.0);
        this->declare_parameter("white_threshold", 2000);
        this->declare_parameter("us_stop_dist", 0.065);
        this->declare_parameter("line_kp", -0.0005);
        this->declare_parameter("line_angular_limit", 0.3);

        // Python: self.goal_x = self.get_parameter('goal_x').value
        // C++:    goal_x_ = this->get_parameter("goal_x").as_double();
        goal_x_ = this->get_parameter("goal_x").as_double();
        goal_y_ = this->get_parameter("goal_y").as_double();
        white_threshold_ = this->get_parameter("white_threshold").as_int();
        us_stop_dist_ = this->get_parameter("us_stop_dist").as_double();
        line_kp_ = this->get_parameter("line_kp").as_double();
        line_angular_limit_ = this->get_parameter("line_angular_limit").as_double();

        // ----- 퍼블리셔 -----
        // Python: self.cmd_pub = self.create_publisher(Twist, '/pinky1/cmd_vel', 10)
        // C++:    cmd_pub_ = this->create_publisher<타입>("토픽명", 10);
        cmd_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("/pinky1/cmd_vel", 10);
        status_pub_ = this->create_publisher<std_msgs::msg::Bool>("/pinky1/parking/auto_done", 10);

        // ----- 액션 클라이언트 -----
        // Python: ActionClient(self, NavigateToPose, '/pinky1/navigate_to_pose')
        nav_client_ = rclcpp_action::create_client<NavigateToPose>(this, "/pinky1/navigate_to_pose");

        // ----- 서브스크립션 -----
        // Python: self.create_subscription(PoseWithCovarianceStamped, '/pinky1/amcl_pose', ...)
        // C++:    this->create_subscription<타입>("토픽명", QoS, 콜백);
        amcl_sub_ = this->create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
            "/pinky1/amcl_pose",
            rclcpp::QoS(1).reliable().transient_local(),
            std::bind(&AutoParkingNode::pose_callback, this, std::placeholders::_1));

        ir_sub_ = this->create_subscription<std_msgs::msg::UInt16MultiArray>(
            "/pinky1/ir_sensor/range", 10,
            std::bind(&AutoParkingNode::ir_callback, this, std::placeholders::_1));

        us_sub_ = this->create_subscription<sensor_msgs::msg::Range>(
            "/pinky1/us_sensor/range", 10,
            std::bind(&AutoParkingNode::us_callback, this, std::placeholders::_1));

        start_sub_ = this->create_subscription<std_msgs::msg::Bool>(
            "/pinky1/parking/auto_start", 10,
            std::bind(&AutoParkingNode::start_callback, this, std::placeholders::_1));

        // ----- 타이머 -----
        // Python: self.create_timer(0.1, self._rotation_tick)
        // C++:    this->create_wall_timer(100ms, 콜백);
        rotation_timer_ = this->create_wall_timer(
            100ms, std::bind(&AutoParkingNode::rotation_tick, this));

        nav_check_timer_ = this->create_wall_timer(
            1s, std::bind(&AutoParkingNode::check_nav_server, this));

        RCLCPP_INFO(this->get_logger(), "AutoParkingNode 시작. 목표=(%f, %f)", goal_x_, goal_y_);
    }

private:
    // =================================================
    // Phase 1: Nav2 이동 (구현 완료)
    // =================================================

    void check_nav_server()
    {
        if (nav_client_->action_server_is_ready()) {
            RCLCPP_INFO(this->get_logger(), "Nav2 서버 준비 완료. 이동 시작.");
            nav_check_timer_->cancel();
            started_ = true;
            go_home();
        }
    }

    void go_home()
    {
        auto goal = NavigateToPose::Goal();
        goal.pose.header.frame_id = "map";
        goal.pose.header.stamp = this->now();
        goal.pose.pose.position.x = goal_x_;
        goal.pose.pose.position.y = goal_y_;

        RCLCPP_INFO(this->get_logger(), "목표 전송: (%f, %f)", goal_x_, goal_y_);

        // Python: future = self._nav_client.send_goal_async(goal, feedback_callback=...)
        //         future.add_done_callback(self._nav_goal_response)
        // C++:    SendGoalOptions에 콜백들을 묶어서 한번에 보냄
        auto options = rclcpp_action::Client<NavigateToPose>::SendGoalOptions();
        options.feedback_callback =
            std::bind(&AutoParkingNode::nav_feedback, this,
                      std::placeholders::_1, std::placeholders::_2);
        options.result_callback =
            std::bind(&AutoParkingNode::nav_result, this, std::placeholders::_1);

        nav_client_->async_send_goal(goal, options);
    }

    // Python: def _nav_feedback(self, feedback_msg):
    //             dist = feedback_msg.feedback.distance_remaining
    void nav_feedback(
        GoalHandleNav::SharedPtr,
        const std::shared_ptr<const NavigateToPose::Feedback> feedback)
    {
        RCLCPP_INFO_THROTTLE(this->get_logger(), *this->get_clock(), 1000,
            "이동 중... 남은 거리: %.2f m", feedback->distance_remaining);
    }

    // Python: def _nav_result(self, future):
    //             status = future.result().status
    void nav_result(const GoalHandleNav::WrappedResult & result)
    {
        if (result.code == rclcpp_action::ResultCode::SUCCEEDED) {
            RCLCPP_INFO(this->get_logger(), "도착. IR 주차 모드 시작.");
            start_ir_parking();
        } else {
            RCLCPP_ERROR(this->get_logger(), "네비게이션 실패. 종료.");
            finish(false);
        }
    }

    // =================================================
    // Phase 2: IR 라인트레이싱  ← TODO: 같이 채울 부분
    // =================================================

    void start_ir_parking()
    {
        ir_active_ = true;
    }

    void ir_callback(const std_msgs::msg::UInt16MultiArray::SharedPtr /*msg*/)
    {
        // TODO: Python의 _ir_callback 로직을 여기에 구현
    }

    // =================================================
    // Phase 3: 초음파 정밀 주차  ← TODO: 같이 채울 부분
    // =================================================

    void start_us_parking()
    {
        us_active_ = true;
    }

    void us_callback(const sensor_msgs::msg::Range::SharedPtr /*msg*/)
    {
        // TODO: Python의 _us_callback 로직을 여기에 구현
    }

    // =================================================
    // Phase 4: 180도 회전  ← TODO: 같이 채울 부분
    // =================================================

    void start_rotation()
    {
        rotating_ = true;
        target_yaw_set_ = false;
    }

    void rotation_tick()
    {
        // TODO: Python의 _rotation_tick 로직을 여기에 구현
    }

    // =================================================
    // 공통 콜백
    // =================================================

    // Python: def _pose_callback(self, msg):
    //             self.current_yaw = quat_to_yaw(msg.pose.pose.orientation)
    void pose_callback(const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
    {
        auto & q = msg->pose.pose.orientation;
        current_yaw_ = std::atan2(
            2.0 * (q.w * q.z + q.x * q.y),
            1.0 - 2.0 * (q.y * q.y + q.z * q.z));
        yaw_valid_ = true;
    }

    void start_callback(const std_msgs::msg::Bool::SharedPtr msg)
    {
        if (msg->data && !started_) {
            started_ = true;
            RCLCPP_INFO(this->get_logger(), "주차 진입 시작 신호 수신.");
            go_home();
        }
    }

    // Python: def _finish(self, success):
    //             self.status_pub.publish(Bool(data=success))
    void finish(bool success)
    {
        auto msg = std_msgs::msg::Bool();
        msg.data = success;
        status_pub_->publish(msg);
        done_ = true;
        rclcpp::shutdown();
    }

    // =================================================
    // 멤버 변수 (Python에서 self.xxx 였던 것들)
    // =================================================

    // 상수
    static constexpr double LINEAR_VEL = 0.05;
    static constexpr double ANGULAR_VEL = 0.5;
    static constexpr double ROTATION_YAW_TOLERANCE = 3.0 * M_PI / 180.0;

    // 파라미터
    double goal_x_{}, goal_y_{};
    int white_threshold_{};
    double us_stop_dist_{};
    double line_kp_{};
    double line_angular_limit_{};

    // 상태
    bool started_ = false;
    bool ir_active_ = false;
    bool us_active_ = false;
    bool rotating_ = false;
    bool done_ = false;
    bool yaw_valid_ = false;
    bool target_yaw_set_ = false;
    double current_yaw_ = 0.0;
    double target_yaw_ = 0.0;

    // ROS 인터페이스
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_pub_;
    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr status_pub_;
    rclcpp_action::Client<NavigateToPose>::SharedPtr nav_client_;
    rclcpp::Subscription<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr amcl_sub_;
    rclcpp::Subscription<std_msgs::msg::UInt16MultiArray>::SharedPtr ir_sub_;
    rclcpp::Subscription<sensor_msgs::msg::Range>::SharedPtr us_sub_;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr start_sub_;
    rclcpp::TimerBase::SharedPtr rotation_timer_;
    rclcpp::TimerBase::SharedPtr nav_check_timer_;
};

// =================================================
// Python: def main(args=None):
//             rclpy.init(args=args)
//             node = AutoParkingNode()
//             rclpy.spin(node)
// =================================================
int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<AutoParkingNode>());
    rclcpp::shutdown();
    return 0;
}
